doc/proposals/scorecard-custom-tests.md

```diff
@@ -0,0 +1,403 @@
+
+Test setup steps for a test of this type can be defined using YAML such as:
+```
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: example-deployment
+  labels:
+    app: nginx
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      app: nginx
+...
+```
+
+#### Complex Test Format
+Complex tests include logic that is very specific or complex in nature to a given operator.  For example, a database operator that might want to test backup and restore operations that might involve multiple CRs and multiple Kube resources.  
+To support this type of test, test logic would be stored within a container image, then the test  essentially becomes the execution of that test container image.  You would then create a test assertion that checks the test pod’s status.   The scorecard would read the Complex test Pod’s log for more detailed information that could be presented to the end-user.
+
+In this scenario, a test setup step would look as follows:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: custom-test-of-something
+  namespace: mynamespace
+spec:
+  restartPolicy: OnFailure
+  containers:
+  - name: mytest
+    image: someuser/sometestimage:3.6.9-ubi8
+    Env:
+…
+```
+
+Then the test assertion might look as follows:
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+ name: custom-test-of-something
+status:
+ phase: Successful
+```
+
+#### Static Check Test Format
+
+Custom tests that are merely a static check are implemented in a similar fashion as a Complex test, however, they do not require a Kube connection necessarily.  A static check custom test would be started as a Pod, with any required on-disk manifests bundled within a ConfigMap that is available to the static check test Pod.  A test assertion would be used to determine if the static check Pod was successful or not just like the other custom test types.  Static check test logs are contained within the Pod’s log, the scorecard would read the Pod’s logs to display back to the end user.
+
+### Custom Test Assertion Output Format
+
+With all scorecard test types, the test assertion output is translated into a scorecard v1alpha2 ScorecardTestResult structure for consistency in reporting test results.
+
+### Custom Test Packaging and Configuration
+
+In this design proposal, the test assertions are YAML files, test steps are YAML files as well.  Container images holding the custom test logic are expected to be pulled from a registry that is accessible in the Kube test environment that is executing the scorecard tests.
+
+Custom tests and the YAML files that make them up would be accessible via URL or from a local disk location.  Custom test locations would be configured from within the scorecard configuration file.  For example, the scorecard configuration might be as follows to indicate the presence of custom tests from different sources:
+```
+scorecard:
+  # Setting a global scorecard option
+  output: json
+  plugins:
+    - tests:
+        init-timeout: 60
```

Comment:
We may want to think about supporting resource limits here as well. Tests could easily gobble up significant cluster resources when exercising the Operator. It should be possible to express limits here and have that be applied. The question is: would it be enough to apply them to the namespace where the tests take place?

Reply:
I added an example and a note that resource limits could be added into the scorecard config and then applied to tests (e.g. Pods).  I'm not sure if its too aggressive to apply limits on the entire namespace but certainly at the test/pod level.